test/lit/passes/gsi-nontype.wast

```diff
@@ -267,3 +267,61 @@
   )
 )
 
+;; Nested struct.gets that seem optimizable.
+(module
+  (rec
+    ;; CHECK:      (rec
+    ;; CHECK-NEXT:  (type $func (func (result anyref)))
+    (type $func (func (result anyref)))
+    ;; CHECK:       (type $outer (sub (struct (field (ref $inner)))))
+    (type $outer (sub (struct (field (ref $inner)))))
+    ;; CHECK:       (type $inner (sub (struct (field (ref $func)))))
+    (type $inner (sub (struct (field (ref $func)))))
+  )
+
+  ;; CHECK:      (type $3 (func (result anyref)))
+
+  ;; CHECK:      (global $global.unnested.0 (ref (exact $inner)) (struct.new $inner
+  ;; CHECK-NEXT:  (ref.func $func)
+  ;; CHECK-NEXT: ))
+
+  ;; CHECK:      (global $global (ref $outer) (struct.new $outer
+  ;; CHECK-NEXT:  (global.get $global.unnested.0)
+  ;; CHECK-NEXT: ))
+  (global $global (ref $outer) (struct.new $outer
+    (struct.new $inner
+      (ref.func $func)
+    )
+  ))
+
+  ;; CHECK:      (func $func (type $func) (result anyref)
+  ;; CHECK-NEXT:  (unreachable)
+  ;; CHECK-NEXT: )
+  (func $func (type $func) (result anyref)
+    (unreachable)
+  )
+
+  ;; CHECK:      (func $caller (type $3) (result anyref)
+  ;; CHECK-NEXT:  (call_ref $func
+  ;; CHECK-NEXT:   (struct.get $inner 0
+  ;; CHECK-NEXT:    (global.get $global.unnested.0)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $caller (result anyref)
+    (call_ref $func
+      ;; TODO: If we did two passes, we could optimize this one too.
```

Comment:
Why doesn't this get optimized via un-nesting as well?

Reply:
It would just take another pass. Maybe worth it as the TODO says.

(We could do it in a single pass with more complexity, or if we traversed in the other direction, in theory...)

Reply:
Oh, we can only do one level of un-nesting per pass because we're function-parallel and therefore cannot actually install the un-nested globals before we move on to the parent? That seems worth fixing! There's no reason why the globals wouldn't have very deep initialization trees, and we don't want to have to run this pass many times to fully optimize them.

Reply:
Yeah, it does seem worth fixing. We could perhaps just take a lock around global creation.